File: Sujin/Ch14_BinaryTree&BinarySearchTree/783_Minimum_Distance_Between_BST_Nodes.cpp
class Solution {
public:
    vector<int> v;
    
    void DFS(TreeNode* root){
        if(!root) return;
        if(root) v.push_back(root->val);
        
        DFS(root->left);
        DFS(root->right);
    }
    
    int minDiffInBST(TreeNode* root) {
        DFS(root);
        
        sort(v.begin(), v.end());
        vector<int>::iterator it;
        int gap = 100001;
        int tmp = v[0];
        
        for(it=v.begin(); it!=v.end()-1; it++){
            int tmp_gap = abs(*it-*(it+1));
            gap = min(gap, tmp_gap); // sliding window 이용
        }
        return gap;
    }
};

File: Suyeon/ch14_Tree/783_Minimum_Distance_Between_BST_Nodes.py
# Definition for a binary tree node.
import sys

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        min_value = sys.maxsize
        before = sys.maxsize  # 이전값 저장

        def travel(root):
            nonlocal min_value
            nonlocal before
            if root is None:
                return

            travel(root.left)
            min_value = min(abs(before-root.val) , min_value)  # 현재 값과 이전값의 차로 최소값 갱신 
            before = root.val  # 이전값으로 갱신
            travel(root.right)

        travel(root)

        return min_value